python/tank/log.py

```diff
@@ -235,6 +237,125 @@ class LogManager(object):
+            # we will try to rename it first. If that fails right away as expected,
+            # we don't try any rollover and append to the current log file.
+            # and raise the _disable_rollover flag.
+
+            try:
+                os.rename(self.baseFilename, temp_backup_name)
+            except:
+                # It failed, so we'll simply append from now on.
+                log.debug("Can't rename main file:", exc_info=True)
+                self._handle_rename_failure("a", disable_rollover=True)
+                return
+
+            # Everything went well, so now simply move the log file back into place
+            # so doRollover can do its work.
+            try:
+                os.rename(temp_backup_name, self.baseFilename)
+            except:
+                # For some reason we couldn't move the backup in its place.
+                log.debug(
+                    "Unexpected issue renaming '%s' to '%s':" % (temp_backup_name, self.baseFilename),
+                    exc_info=True
+                )
+                # The main log file doesn't exist anymore, so create a new file.
+                # Don't disable the rollover, this has nothing to do with rollover
+                # failing.
+                self._handle_rename_failure("w")
+                return
+
+            # Now, that we are back in the original state we were in,
+            # were pretty confident that the rollover will work. However, due to
+            # any number of reasons it could still fail. If it does, simply
+            # disable rollover and append to the current log.
+            try:
+                super(LogManager._SafeRotatingFileHandler, self).doRollover()
+            except:
+                # Something probably failed trying to rollover the backups,
+                # since the code above proved that in theory the main log file
+                # should be renamable. In any case, we didn't succeed in renaming,
+                # so disable rollover and reopen the main log file in append mode.
+                log.debug("Rollover failed:", exc_info=True)
+                self._handle_rename_failure("a", disable_rollover=True)
+
+        def _handle_rename_failure(self, mode, disable_rollover=False):
+            """
+            Reopen the log file in the specific mode and optionally disable
+            future rollover operations.
+
+            :param str mode: Mode in which to reopen the main log file.
+            :param bool disable_rollover: If True, rollover won't be possible in the
+                future. Defaults to False.
+            """
+            # Keep track that the rollover failed.
+            self._disable_rollover = disable_rollover
+            # If the file has been closed, reopen it in append mode.
+            if not self.stream:
+                self.mode = mode
+                self.stream = self._open()
+
+        def shouldRollover(self, record):
```

Comment:
does that mean that once it fails to roll over, it never tries again? Maybe that's worth adding in the log messages too - "willl not attempt to roll over the log any more times during this tk session" - or something like that?